nixos/modules/services/networking/vpp.nix

```diff
@@ -0,0 +1,543 @@
+    ;
+  cfg = config.services.vpp;
+
+  # Converts the settings option to a string
+  parseConfig =
+    cfg:
+    let
+      inherit (lib.lists) sort concatMap filter;
+
+      sortedAttrs =
+        set:
+        sort (
+          l: r:
+          if l == "extraConfig" then
+            false # Always put extraConfig last
+          else if builtins.isAttrs set.${l} == builtins.isAttrs set.${r} then
+            l < r
+          else
+            builtins.isAttrs set.${r} # Attrsets should be last, makes for a nice config
+          # This last case occurs when any side (but not both) is an attrset
+          # The order of these is correct when the attrset is on the right
+          # which we're just returning
+        ) (builtins.attrNames set);
+
+      # Specifies an attrset that encodes the value according to its type
+      encode =
+        name: value:
+        {
+          null = [ ];
+          bool = lib.optional value name;
+          int = [ "${name} ${builtins.toString value}" ];
+
+          # extraConfig should be inserted verbatim
+          string = [ (if name == "extraConfig" then value else "${name} ${value}") ];
+
+          # Values like `foo = [ "bar" "baz" ];` should be transformed into
+          #   foo bar
+          #   foo baz
+          list = concatMap (encode name) value;
+
+          # Values like `foo = { bar = { baz = true; }; bar2 = { baz2 = true; }; };` should be transformed into
+          #   foo bar {
+          #     baz
+          #   }
+          #   foo bar2 {
+          #     baz2
+          #   }
+          set = concatMap (
+            subname:
+            lib.optionals (value.${subname} != null) (
+              [ "${name} ${subname} {" ] ++ (map (line: "  ${line}") (toLines value.${subname})) ++ [ "}" ]
+            )
+          ) (filter (v: v != null) (builtins.attrNames value));
+        }
+        .${builtins.typeOf value};
+
+      # One level "above" encode, acts upon a set and uses encode on each name,value pair
+      toLines = set: concatMap (name: encode name set.${name}) (sortedAttrs set);
+
+      # Moves top-level attrsets into a dummy "" value, so that `section = {...}` is transformed to `section  {...}`
+      parseSections =
+        set:
+        builtins.mapAttrs (name: value: {
+          "" = value;
+        }) set;
+    in
+    lib.strings.concatStringsSep "\n" (toLines (parseSections cfg));
+
+  semanticTypes = with types; rec {
+    vppAtom = nullOr (oneOf [
+      int
+      bool
+      str
+    ]);
+    vppAttr = attrsOf vppAll;
+    vppAll =
+      (oneOf [
+        vppAtom
+        (listOf vppAtom)
+        vppAttr
+      ])
+      // {
+        # Since this is a recursive type and the description by default contains
+        # the description of its subtypes, infinite recursion would occur without
+        # explicitly breaking this cycle
+        description = "vpp values (atoms (null, str, int, bool), list of atoms, or attrsets of vpp values)";
+      };
+  };
```

Comment:
You may have taken this into account already, but have you checked `pkgs.formats.*` to see if there's a compatible config format already available? I'm not sure if all formats have `toFormat` functions, but I thought it was worth pointing out.
For reference: https://github.com/NixOS/nixpkgs/blob/master/nixos%2Fmodules%2Fservices%2Fweb-apps%2Fdashy.nix#L154

Reply:
I have, VPP seems to use a completely custom config format that's not even standardized (nevermind available in `pkgs.formats`). If you find a way to make it work that'd be great, but I think we're stuck with the custom parser.